File: R/d09.r

input = strsplit(readLines("data/d09.txt", warn = FALSE), " ")

#' @export 
solve = function(input, return_knots) {
    len = sum(sapply(input, function(x) as.integer(x[2])))
    ins = vector("integer", len)
    pos = 1

    # parse position of lead knot throughout
    for (moves in input) {
        dir = moves[1]
        int = as.integer(moves[2])
        ins[pos:(pos+int)] = dir
        pos = pos + int
    }

    pos = 1
    hpos = tpos = vector("list", len)
    tpos = c(tpos, list())
    hx=hy=0

    for (ii in seq_len(length(ins)-1)) {
        if (ins[ii] == "R") hx = hx + 1
        if (ins[ii] == "L") hx = hx - 1
        if (ins[ii] == "U") hy = hy + 1
        if (ins[ii] == "D") hy = hy - 1
        hpos[[ii]] = c(hx, hy)
    }

    ans = c()

    # loop through knots offsetting position against predecessor
    for (knot in 1:9) {
        if (knot != 1) hpos = tpos
        tpos[[1]] = c(0,0)
        for (ii in seq_len(length(hpos))) {
            difs = hpos[[ii]] - tpos[[ii]]
            if (any(abs(difs) > 1)) {
                if (difs[1] > 0 & difs[2] == 0) {
                    tpos[[ii]] = tpos[[ii+1]] = tpos[[ii]] + c(1,0)
                } else if (difs[1] > 0 & difs[2] > 0) {
                    tpos[[ii]] = tpos[[ii+1]] = tpos[[ii]] + c(1,1)
                } else if (difs[1] > 0 & difs[2] < 0) {
                    tpos[[ii]] = tpos[[ii+1]] = tpos[[ii]] +  c(1,-1)
                } else if (difs[1] < 0 & difs[2] == 0) {
                    tpos[[ii]] = tpos[[ii+1]] = tpos[[ii]] + c(-1,0)
                } else if (difs[1] < 0 & difs[2] < 0) {
                    tpos[[ii]] = tpos[[ii+1]] = tpos[[ii]] + c(-1,-1)
                } else if (difs[1] < 0 & difs[2] > 0) {
                    tpos[[ii]] = tpos[[ii+1]] = tpos[[ii]] + c(-1,1)
                } else if (difs[1] == 0 & difs[2] >= 1) {
                    tpos[[ii]] = tpos[[ii+1]] = tpos[[ii]] + c(0,1)
                } else if (difs[1] == 0 & difs[2] <= -1) {
                    tpos[[ii]] = tpos[[ii+1]] = tpos[[ii]] + c(0,-1)
                }
            } else {
                tpos[[ii+1]] = tpos[[ii]]
            }
        }
        if (knot %in% return_knots) {
            ans = c(ans, nrow(unique(do.call(rbind, tpos))))
        }
    }
    return(ans)
}

File: Python/d09.py

class Knots:
    """
    Solve day 9 Aoc 2022
    """

    def __init__(self, input: list[tuple[str, int]], n_knots: int):
        self.input = input
        self.n_knots = n_knots
        self.hpos = self.first_head(input=self.input)
        self.knot_pos = self.get_knots(n_knots=self.n_knots, head=self.hpos)

    def first_head(self, input: list[tuple[int, int]]) -> list[tuple[int, int]]:
        """
        Record the 2 dimensional position of the 'head' of the rope
        at any given point in time.
        """
        res = []
        x = 0
        y = 0
        for mv in input:
            direction, distance = mv
            for _ in range(0, distance):
                if direction == 'R':
                    x += 1
                elif direction == 'L':
                    x -= 1
                elif direction == 'U':
                    y += 1
                elif direction == 'D':
                    y -= 1
                res.append((x, y))
        return res

    def get_knots(self, n_knots: int, head: list[tuple[int, int]]) -> list[tuple[int, int]]:
        """
        Record the 2 dimensional position of any 'knot' in the rope
        at any given point in time.
        """
        knots = []
        for _ in range(0, n_knots):
            knot = [(0, 0)]
            for position in head:
                tail_x, tail_y = knot[len(knot)-1]
                head_x, head_y = position
                while abs(head_x - tail_x) > 1 or abs(head_y - tail_y) > 1:

                    if (head_x > tail_x) and ((tail_x + 1, tail_y) != (head_x, head_y)):
                        tail_x += 1
                    elif (head_x < tail_x) and ((tail_x - 1, tail_y) != (head_x, head_y)):
                        tail_x -= 1

                    if (head_y > tail_y) and ((tail_x, tail_y + 1) != (head_x, head_y)):
                        tail_y += 1
                    elif (head_y < tail_y) and ((tail_x, tail_y - 1) != (head_x, head_y)):
                        tail_y -= 1

                    knot.append((tail_x, tail_y))

            knots.append(knot)
            head = knots[len(knots)-1]
        return knots


with open("data/d09.txt") as f:
    input = f.read().split("\n")

dat = []

for i in input:
    dir, mv = i.split()
    dat.append((dir, int(mv)))

rope = Knots(dat, 9)

# Part 1 Answer
print(len(set(rope.knot_pos[0])))

# Part 2 Answer
print(len(set(rope.knot_pos[8])))
